projects/observability/src/shared/dashboard/widgets/top-n/data/top-n-data-source.model.ts

```diff
@@ -46,41 +34,62 @@ export class TopNDataSourceModel extends GraphQlDataSourceModel<TopNWidgetDataFe
 
   public getData(): Observable<TopNWidgetDataFetcher> {
     return of({
-      scope: this.entityType,
-      getData: (metricSpecification: MetricAggregationSpecification) => this.fetchDataWithMetric(metricSpecification)
+      scope: this.context,
+      getData: (metricSpecification: ExploreSpecification) => this.fetchDataWithMetric(metricSpecification)
     });
   }
 
-  private fetchDataWithMetric(metricSpecification: MetricAggregationSpecification): Observable<TopNWidgetValueData[]> {
-    return this.queryWithNextBatch<EntitiesGraphQlQueryHandlerService, EntitiesResponse>(filters => ({
-      requestType: ENTITIES_GQL_REQUEST,
-      entityType: this.entityType,
+  private readonly specBuilder: ExploreSpecificationBuilder = new ExploreSpecificationBuilder();
+
+  private nameAttributeSpec: ExploreSpecification = this.specBuilder.exploreSpecificationForKey('name');
+  private idAttributeSpec: ExploreSpecification = this.specBuilder.exploreSpecificationForKey('id');
+
+  private fetchDataWithMetric(metricSpecification: ExploreSpecification): Observable<TopNWidgetValueData[]> {
+    return this.queryWithNextBatch<ExploreGraphQlQueryHandlerService, GraphQlExploreResponse>(filters => ({
+      requestType: EXPLORE_GQL_REQUEST,
+      context: this.context,
       limit: this.resultLimit,
       timeRange: this.getTimeRangeOrThrow(),
-      properties: [this.attributeSpecification, metricSpecification],
+      selections: [this.nameAttributeSpec, this.idAttributeSpec, metricSpecification],
       filters: filters,
-      sort: {
-        direction: 'DESC',
-        key: metricSpecification
-      }
+      groupBy: {
+        keys: [this.nameAttributeSpec.name, this.idAttributeSpec.name]
+      },
+      orderBy: [
+        {
+          direction: 'DESC',
+          key: metricSpecification
+        }
+      ]
     })).pipe(
       map(response =>
         response.results.map(entity => ({
-          label: entity[this.attributeSpecification.resultAlias()] as string,
-          value: (entity[metricSpecification.resultAlias()] as MetricAggregation).value,
+          label: (entity[this.nameAttributeSpec.resultAlias()] as GraphQlExploreResultValue).value as string,
+          value: (entity[metricSpecification.resultAlias()] as GraphQlExploreResultValue).value as number,
           entity: {
-            [entityIdKey]: entity[entityIdKey],
-            [entityTypeKey]: this.entityType
+            [entityIdKey]: (entity[this.idAttributeSpec.resultAlias()] as GraphQlExploreResultValue).value as string,
+            [entityTypeKey]: this.getEntityTypeForContext(this.context)
           }
         }))
       )
     );
   }
+
+  private getEntityTypeForContext(context: string): string {
+    switch (context) {
+      case 'API':
```

Comment:
currently we do not have an enum for explore context values. Should I add one or is there a reason why we haven't added it yet? @skjindal93 @aaron-steinfeld

Reply:
Shouldn't need a new one - explorer context is just a superset of other enums - trace type, entity type etc. Because of that, the translation here isn't needed - these should be identical.

Reply:
My comment was a little premature - this falls under the same family as the other comments - doesn't make sense except for entities.

Reply:
I'm working with the assumption that top-n will want to fetch entity information ultimately. I don't think its possible / a worthwhile pursuit to write something thats super generic without complicating it.

Here, the return type is TopnData with entity information.

Reply:
Agree it's not worth doing right now - let's just adjust this one to have a name + api that make sense specifically for entities.